File: C++/203. Remove Linked List Elements.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* removeElements(ListNode* head, int val) {
        ListNode* dummy_head = new ListNode(0);
        dummy_head->next = head;
        ListNode* p = head;
        ListNode* pre = dummy_head;
        while(p != NULL){
            if (p->val == val){
                pre->next = p->next;
                p = p->next;
            }
            else{
                pre = p;
                p = p->next;
            }
        }
        return dummy_head->next;
    }
};

File: 5-Problem-A-Day/203. Remove Linked List Elements.java
/**
 * Definition for singly-linked list. public class ListNode { int val; ListNode next; ListNode(int
 * x) { val = x; } }
 */
class Solution {
    public ListNode removeElements(ListNode head, int val) {
        ListNode dummy = new ListNode(0), prev = dummy;
        dummy.next = head;
        while (head != null) {
            if (head.val == val)
                prev.next = head.next;
            else
                prev = prev.next;
            head = head.next;
        }
        return dummy.next;
    }
}
